### PR TITLE
Ipreservation newendpoints updated

### PR DESCRIPTION
**Pull request by AniJ98**
## 📝 Description

This PR introduces comprehensive functionality and tests for Reserved IP Addresses in the Linodego client. The changes are necessary to provide robust support for managing Reserved IP addresses through the Linode API. The changes include:
1. Implementation of core Reserved IP operations:
* Listing Reserved IPs
* Getting a specific Reserved IP
* Reserving a new IP
* Deleting a Reserved IP
2. Test coverage:
* TestReservedIPAddresses_InsufficientPermissions: Verifies behavior when the user lacks necessary permissions.
* TestReservedIPAddresses_EndToEndTest: Performs a complete workflow test for users with proper permissions.
* TestReservedIPAddresses_ListIPAddressesVariants: Tests various scenarios for listing IP addresses.
* TestReservedIPAddresses_GetIPAddressVariants: Covers different cases for retrieving a specific IP address.
* TestReservedIPAddresses_ReserveIPVariants: Tests multiple scenarios for reserving an IP, including edge cases.
* TestReservedIPAddresses_DeleteIPAddressVariants: Verifies deletion functionality under various conditions.

## ✔️ How to Test

**What are the steps to reproduce the issue or verify the changes?**

What are the steps to reproduce the issue or verify the changes?
To verify the changes related to Reserved IP functionality, follow these steps:
1. Ensure you have a valid Linode API token with the "can_reserve_ip" permission enabled. Set up your environment: 
`export LINODE_TOKEN="your_token_here"` 
Additionally you need to set the following if you want to test it in a different environment: 
`export LINODE_URL="https://api.linode.com"`
`export LINODE_API_VERSION="v4beta"`

2. Navigate to the test directory within the linodego project. Update the LINODE_TOKEN in the Makefile: `LINODE_TOKEN="your_token_here"`

3. Run the tests using one of the following commands: To run all integration tests: 
`make testint`
To run only Reserved IP related tests: 
`go test -v ./integration -run TestReservedIPAddresses `
To run a specific test:
`go test -v ./integration -run TestReservedIPAddresses_EndToEndTest`

Verify the test output for any failures or unexpected behavior.

Note:
* Ensure you have proper permissions and sufficient quota in your Linode account to perform Reserved IP operations. Some tests may create and delete resources, so use a testing environment if possible.

### CODE REVIEW

File: test/integration/network_reserved_ips_test.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+	os.Setenv("LINODE_TOKEN", dummyToken)
+
+	client, teardown := createTestClient(t, "fixtures/TestReservedIpAddresses_InsufficientPermissions")
+	defer teardown()
+	defer os.Setenv("LINODE_TOKEN", original)
+
+	filter := ""
+	i, getReservedIpsError := client.GetReservedIPs(context.Background(), NewListOptions(0, filter))
+	if getReservedIpsError != nil {
+		t.Logf("Error listing ipaddresses, expected struct, got error %v", getReservedIpsError)
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: Don't you want to fail the test if you hit an error instead of continuing to execute the rest of the test case?

File: test/integration/network_reserved_ips_test.go [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	}
+
+	if len(i) == 0 {
+		t.Logf("Expected a list of ipaddresses, but got none %v", i)
+	}
+
+	// Attempt to reserve an IP address
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go:35]: You expect this to fail, right? You should probably fail the test in the case you get a positive result.

File: test/integration/network_reserved_ips_test.go [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+	}
+
+	// Attempt to get a reserved IP address
+	address := "172.28.3.4"
+	ipaddress, getReservedIpError := client.GetReservedIPAddress(context.Background(), address)
+	if getReservedIpError != nil {
+		t.Logf("Error getting ipaddress, expected struct, got %v and error %v", ipaddress, getReservedIpError)
+	}
+
+	// Attempt to delete a reserved IP address
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go:54]: Same as the reserve comment.

File: test/integration/network_reserved_ips_test.go [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+		t.Logf("Failed to reserve IP: %v", reserveIpError)
+	} else {
+		t.Logf("Successfully reserved IP: %+v", reservedIp)
+	}
+
+	// Attempt to get a reserved IP address
+	address := "172.28.3.4"
+	ipaddress, getReservedIpError := client.GetReservedIPAddress(context.Background(), address)
+	if getReservedIpError != nil {
+		t.Logf("Error getting ipaddress, expected struct, got %v and error %v", ipaddress, getReservedIpError)
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: Should probably fail the test here.

File: test/integration/network_reserved_ips_test.go [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+	"context"
+	"os"
+	"testing"
+
+	. "github.com/linode/linodego"
+)
+
+// TestReservedIPAddresses_InsufficientPermissions tests the behavior when a user account
+// doesn't have the can_reserve_ip flag enabled
+func TestReservedIPAddresses_InsufficientPermissions(t *testing.T) {
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go:14]: Part of the test checks for insufficient permission case (reserve / delete) while the rest of the cases covered here are sanity checks that could probably be moved to another test case.

File: test/integration/network_reserved_ips_test.go [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+	}
+}
+
+// TestReservedIPAddresses_ListIPAddressesVariants tests various scenarios for listing IP addresses
+func TestReservedIPAddresses_ListIPAddressesVariants(t *testing.T) {
+	client, teardown := createTestClient(t, "fixtures/TestReservedIPAddresses_ListIPAddressesVariants")
+	defer teardown()
+
+	filter := ""
+	ipList, listIpErr := client.GetReservedIPs(context.Background(), NewListOptions(0, filter))
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: Can't easily tell what the expected result is. Looking at `fixtures/TestReservedIPAddresses_ListIPAddressesVariants` an empty list is expected, and all other cases should be considered test failures using t.Fatalf

**Review by lalsheni — CHANGES_REQUESTED** [review]
I think the tests need be updated to fail in the unexpected cases.

File: test/integration/network_reserved_ips_test.go (hunk 1 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: Yes I have now changed the test to fail if the expected error does not occur.

File: test/integration/network_reserved_ips_test.go (hunk 2 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go:35]: Yes I have now used errorf() insetad of logf to fail the test in case the expected error does not occur.

File: test/integration/network_reserved_ips_test.go (hunk 3 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go:54]: Made similar changes to fail the test.

File: test/integration/network_reserved_ips_test.go (hunk 4 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: The test now fails if the expected error does not occur.

File: test/integration/network_reserved_ips_test.go (hunk 5 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go:14]: I am pretty sure I checked this, but last I remember we needed permission to list and fetch reserved IP's as well.

Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go:14]: I can confirm that and change it if we do not need permission for them.

File: test/integration/network_reserved_ips_test.go (hunk 6 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: This test is now updated to list reserved IP's belonging to a particular region using filters.

File: network_reserved_ips.go [hunk 7]
```diff
@@ -10,7 +10,7 @@ type ReserveIPOptions struct {
 }
 
 // GetReservedIPs retrieves a list of reserved IP addresses
```
Comment (lalsheni) [inline network_reserved_ips.go]: Update comment to be consistent with name change.

File: test/integration/network_reserved_ips_test.go [hunk 8]
```diff
@@ -2,150 +2,199 @@ package integration
...
 	}
 
-	if reserveIP != nil {
-		// Fetch the reserved IP
-		reservedIP, fetchIpErr := client.GetReservedIPAddress(context.Background(), reserveIP.Address)
-		if fetchIpErr != nil {
-			t.Logf("Error getting ipaddress, expected struct, got %v and error %v", reservedIP, fetchIpErr)
-		}
+	if resIP == nil {
+		t.Fatalf("Reserved IP is nil")
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: Maybe state what the expected result in this case is, and why nil is a failure?

File: test/integration/network_reserved_ips_test.go (hunk 5 above)
Comment (lalsheni) [reply test/integration/network_reserved_ips_test.go:14]: My comment was based on the requirement for the `can_reserve_ip ` which only applies to reserve/unreserve.

Iit seems like you're checking for general permissions to use the feature, which is a different flag that would apply to all new endpoints, but shouldn't matter for this test case. I would just update the comment to reflect `permission to use the reservedIP feature`

File: test/integration/network_reserved_ips_test.go [hunk 9]
```diff
@@ -2,150 +2,199 @@ package integration
...
+	}
 
-		// Verify the IP has been deleted
-		if reserveIP != nil {
-			verifyDeletedIP, fetchDeletedIpErr := client.GetReservedIPAddress(context.Background(), reserveIP.Address)
-			if fetchDeletedIpErr != nil {
-				t.Logf("Error getting ipaddress, expected struct, got %v and error %v", verifyDeletedIP, fetchDeletedIpErr)
-			}
+	if fetchedIP == nil {
+		t.Errorf("Retrieved reserved IP is nil")
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: It's helpful to always include the expected state in the error messages since it simplifies debugging in the future.

File: test/integration/network_reserved_ips_test.go [hunk 10]
```diff
@@ -2,150 +2,199 @@ package integration
...
+	// Delete the reserved IP
+	delErr := client.DeleteReservedIPAddress(context.Background(), resIP.Address)
+	if delErr != nil {
+		t.Fatalf("Error deleting reserved IP address: %v", delErr)
+	}
+
+	// Verify the IP has been deleted
+	_, fetchDelErr := client.GetReservedIPAddress(context.Background(), resIP.Address)
+	if fetchDelErr == nil {
+		t.Errorf("Expected error when fetching deleted IP, got nil")
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: Same as prev. comment about which IP was expected.

File: test/integration/network_reserved_ips_test.go [hunk 11]
```diff
@@ -2,150 +2,199 @@ package integration
...
+	for _, ip := range ipList {
+		if !ip.Reserved {
+			t.Errorf("Expected all IPs to be reserved, but found non-reserved IP: %s", ip.Address)
+		}
+		if ip.Region != "us-east" {
+			t.Errorf("Expected all IPs to be in us-east region, but found IP in %s region: %s", ip.Region, ip.Address)
+		}
+		for _, reservedIP := range reservedIPs {
+			if ip.Address == reservedIP {
+				foundReservedIPs++
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go:190]: I can't find the check for `foundReservedIPs ` == `expected_ips`?

File: test/integration/network_reserved_ips_test.go [hunk 12]
```diff
@@ -2,150 +2,199 @@ package integration
...
 
-	filter := ""
-	ipList, listIpErr := client.GetReservedIPs(context.Background(), NewListOptions(0, filter))
-	if listIpErr != nil {
-		t.Logf("Error listing ipaddresses, expected struct, got error %v", listIpErr)
-	} else {
-		if len(ipList) == 0 {
-			t.Logf("The customer has not reserved an IP %v", ipList)
+	// Reserve two IP addresses in us-east region
+	reservedIPs := make([]string, 2)
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: It's good practice to store constants in a variable since it clarifies its meaning and intention and makes it easier to modify (are all the 2's you used below related to ips?). For example, `expected_ips := 2`.

File: test/integration/network_reserved_ips_test.go [hunk 13]
```diff
@@ -237,48 +306,51 @@ func TestReservedIPAddresses_DeleteIPAddressVariants(t *testing.T) {
...
-	if len(ipList) > 0 {
-		ipToBeDeleted := ipList[len(ipList)-1]
-		deleteErr := client.DeleteReservedIPAddress(context.Background(), ipToBeDeleted.Address)
-		if deleteErr != nil {
-			t.Logf("Error deleting reserved IP address: %v", deleteErr)
-		} else {
-			t.Logf("Successfully deleted reserved IP address: %s", ipToBeDeleted.Address)
-		}
+	if len(ipList) == 0 {
+		t.Skip("No reserved IPs available for testing deletion")
```
Comment (lalsheni) [inline test/integration/network_reserved_ips_test.go]: I think `t.Fatalf` is more appropriate in this case, since you expect one ip to be reserved. 

`Fatalf` wil fail the test right away and skip the rest of the statements in the test unlike `Errorf` which will continue. `Skip` will not mark the test case as failed, but skipped, which is probably not what we want for this case.

File: test/integration/network_reserved_ips_test.go (hunk 8 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: Removed this condition as I dont believe there is a condition where the response would be nil without throwing an error. The error would be caught displayed to the user and test would consequently fail.

File: test/integration/network_reserved_ips_test.go (hunk 5 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go:14]: Yeah that makes more sense. I will be changing the comment.

File: test/integration/network_reserved_ips_test.go (hunk 9 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: I have now included the IP Address which the test failed to retrieve in the error message.

File: test/integration/network_reserved_ips_test.go (hunk 10 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: I have now included the IP that I am trying to fetch unsuccessfully in the error message.

File: test/integration/network_reserved_ips_test.go (hunk 11 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go:190]: I must have missed this check before committing. This check is now present.

File: test/integration/network_reserved_ips_test.go (hunk 12 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: Yes all the 2's are related to the number of IP's that this test will be reserving and then listing. Created a constant instead and used it everywhere.

File: test/integration/network_reserved_ips_test.go (hunk 13 above)
Comment (AniJ98) [reply test/integration/network_reserved_ips_test.go]: Yes we should ideally be failing the test in this case and using t.Fatalf makes more sense in this case. I hvae made the chnage and this particular check now uses t.Fatalf.

**Review by lalsheni — APPROVED** [review]
